sdk/core/AzureCore/Source/Util/CryptoUtil.swift

```diff
@@ -35,62 +35,75 @@ public enum CryptoAlgorithm {
 
     /// Underlying CommonCrypto HMAC algorithm.
     public var hmacAlgorithm: CCHmacAlgorithm {
-        var alg = 0
         switch self {
         case .sha1:
-            alg = kCCHmacAlgSHA1
+            return CCHmacAlgorithm(kCCHmacAlgSHA1)
         case .md5:
-            alg = kCCHmacAlgMD5
+            return CCHmacAlgorithm(kCCHmacAlgMD5)
         case .sha256:
-            alg = kCCHmacAlgSHA256
+            return CCHmacAlgorithm(kCCHmacAlgSHA256)
         case .sha384:
-            alg = kCCHmacAlgSHA384
+            return CCHmacAlgorithm(kCCHmacAlgSHA384)
         case .sha512:
-            alg = kCCHmacAlgSHA512
+            return CCHmacAlgorithm(kCCHmacAlgSHA512)
         case .sha224:
-            alg = kCCHmacAlgSHA224
+            return CCHmacAlgorithm(kCCHmacAlgSHA224)
         }
-        return CCHmacAlgorithm(alg)
     }
 
-    /// Compute a hash of the underlying data using the specfied algorithm.
-    public func hash(_ data: UnsafeRawPointer!, _ len: CC_LONG, _ message: UnsafeMutablePointer<UInt8>!) -> Data {
-        var result: UnsafeMutablePointer<UInt8>
+    /// Digest length for the HMAC algorithm.
+    public var digestLength: Int {
```

Comment:
So, the first thing I did here was write tests and verify that the existing versions were working (caught one bug actually). Then I refactored them to clean things up and to resolve the compiler warnings, and made sure that the tests continued to pass.

The summary of things that changed:
- Simplified implementation of computed properties a little (removed scratch variables)
- Fixed a bug in hash that the tests caught (SHA256 hash was actually calculating SHA224)
- Lifted the core of the HMAC algorithm into the `CryptoAlgorithm` enum, just like we already had done with `hash`
- Fixed the compiler deprecation warnings about `UnsafeMutableBytes`
- Simplified the implementation of `hash` since we never actually needed to use the return value from the `CC_` hash methods
- Got rid of any unnecessary conversion from Data->String or String->Data that was happening in either the hash or the hmac methods

I think that's it!

Reply:
Awesome!